### PR TITLE
feat: Add Selective ATen decompositions

### LINKED ISSUE #2160: ✨[Improvement] Unified + Selective Decomposition System for Dynamo
# Context
Currently, the decomposition system in compile, soon to be integrated into export, is based on the [`core_aten_decompositions`](https://github.com/pytorch/pytorch/blob/6d2162e644152efae91d22208984f1dd2f9bd999/torch/_decomp/__init__.py#L180) set, as here:
https://github.com/pytorch/TensorRT/blob/f7b03f466f62c1027f4025b60a4f5e1f854550a6/py/torch_tensorrt/dynamo/lowering/_decompositions.py#L5
This set of decompositions, while exhaustive, decomposes certain operations which we would like to preserve, including `aten.cudnn_batch_norm` and other batch norm implementations. Such implementations, when combined with utilities for freezing, such as those discussed in #2124, #2128, can greatly improve performance on networks with these operators.

As a bonus, the pre-aot decompositions, [as here](https://github.com/pytorch/TensorRT/tree/f7b03f466f62c1027f4025b60a4f5e1f854550a6/py/torch_tensorrt/dynamo/lowering/substitutions), may be more easily removed/reduced, since it is difficult to integrate these into export currently.

# Proposal
Introduce a new decomposition dictionary which does one of two things:

  **1.** Selectively _remove_ operators from the `core_aten_decompositions` dictionary, to ensure these operators are not decomposed and will stay in their initial state. This allows us to continue using the most up-to-date decompositions dictionary while selectively removing the operators which are preferable to stay as-is.
  **2.** Selectively _keep_ operators in the `core_aten_decompositions` dictionary, to ensure the set of decompositions is fully defined and does not depend on the state of PyTorch's dictionary. This gives more control over the set of decompositions allowed, but requires that modifications are made to the decompositions library more often, since PyTorch may introduce new desirable decompositions often.

### PR DESCRIPTION
**Pull request by gs-olive**
# Description

- Add sets to selectively enable or disable decompositions in Torch
- Add new runtime argument `enable_experimental_decompositions` to enable all core aten decompositions, or a pre-selected subset thereof
- Improve documentation of compilation settings overall

Fixes #2160 

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [ x ] My code follows the style guidelines of this project (You can use the linters)
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ x ] I have made corresponding changes to the documentation
- [ - ] I have added tests to verify my fix or my feature
  - CI verification, documentation changes
- [ x ] New and existing unit tests pass locally with my changes
- [ x ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 1]
```diff
@@ -1,10 +1,187 @@
...
+    aten.zero_,
+    aten.zeros,
+    aten.zeros_like,
+}
+disabled_decompositions: Set[torch._ops.OpOverload] = {}
 
-aten = torch.ops.aten
+TORCH_DECOMPOSITIONS: Dict[torch._ops.OpOverload, Callable] = get_torch_decompositions(
+    enabled_decompositions
+)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py]: Currently, the decompositions are sourced directly from Torch's main registry (via `get_torch_decompositions`) and may not exactly match with the `_core_aten_decompositions`. This is because certain decompositions which we depend on (such as `native_layer_norm`, may occasionally be removed from the core set).

Whenever Torch versions are upgraded, this list should be updated as well.

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 2]
```diff
@@ -1,10 +1,187 @@
...
+TORCH_EXPERIMENTAL_DECOMPOSITIONS: Dict[torch._ops.OpOverload, Callable] = {
+    decomp: _core_aten_decompositions[decomp]
+    for decomp in _core_aten_decompositions
+    if decomp not in disabled_decompositions
+}
+CUSTOM_DECOMPOSITIONS: Dict[torch._ops.OpOverload, Callable] = {}
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py]: The decompositions are three dictionaries:

- **`TORCH_DECOMPOSITIONS`** - the enabled decompositions we've pre-selected
- **`TORCH_EXPERIMENTAL_DECOMPOSITIONS`** - the complete set of `_core_aten_decompositions` Torch provides, minus the set of disabled decompositions. Note that `TORCH_DECOMPOSITIONS` _may not_ be a subset of this set
- **`CUSTOM_DECOMPOSITIONS`** - the decompositions we've written ourselves

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 3]
```diff
@@ -70,12 +247,31 @@ def addmm_replacement(
...
+)
 def reciprocal_replacement(
     input_: torch.Tensor,
 ) -> torch.Tensor:
     return torch.div(1, input_)
 
 
-def get_decompositions():
-    return DECOMPOSITIONS
+def get_decompositions(
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:256]: The dictionary returned by `get_decompositions` is either `TORCH_DECOMPOSITIONS` or `TORCH_EXPERIMENTAL_DECOMPOSITIONS` concatenated with our `CUSTOM_DECOMPOSITIONS`

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 4]
```diff
@@ -1,10 +1,182 @@
...
+    aten.zero_,
+    aten.zeros,
+    aten.zeros_like,
+}
+disabled_decompositions: Set[torch._ops.OpOverload] = {}
 
-aten = torch.ops.aten
+ENABLED_TORCH_DECOMPOSITIONS: Dict[
+    torch._ops.OpOverload, Callable
+] = get_torch_decompositions(enabled_decompositions)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:178]: Currently, the decompositions are sourced directly from Torch's main registry (via `get_torch_decompositions`) and may not exactly match with the `_core_aten_decompositions`. This is because certain decompositions which we depend on (such as `native_layer_norm`, may occasionally be removed from the core set).

Whenever Torch versions are upgraded, this list should be updated as well.

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 5]
```diff
@@ -1,10 +1,182 @@
...
+TORCH_TRT_DECOMPOSITIONS: Dict[torch._ops.OpOverload, Callable] = {}
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:179]: The decompositions are three dictionaries:

- **`ENABLED_TORCH_DECOMPOSITIONS`** - the enabled decompositions we've pre-selected
- **`CORE_ATEN_DECOMPOSITIONS_FILTERED`** (defined in `get_decompositions` below) - the complete set of `_core_aten_decompositions` Torch provides, minus the set of disabled decompositions. Note that `TORCH_DECOMPOSITIONS` _may not_ be a subset of this set
- **`TORCH_TRT_DECOMPOSITIONS`** - the decompositions we've written ourselves

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 6]
```diff
@@ -70,12 +244,33 @@ def addmm_replacement(
...
+)
 def reciprocal_replacement(
     input_: torch.Tensor,
 ) -> torch.Tensor:
     return torch.div(1, input_)
 
 
-def get_decompositions():
-    return DECOMPOSITIONS
+def get_decompositions(
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:256]: The dictionary returned by `get_decompositions` is either `ENABLED_TORCH_DECOMPOSITIONS` or `CORE_ATEN_DECOMPOSITIONS_FILTERED` concatenated with our `TORCH_TRT_DECOMPOSITIONS`

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 7]
```diff
@@ -1,10 +1,182 @@
...
+    get_decompositions as get_torch_decompositions,
+)
 
+aten = torch.ops.aten
 
-DECOMPOSITIONS = {**core_aten_decompositions()}
+_core_aten_decompositions: Dict[
+    torch._ops.OpOverload, Callable
+] = core_aten_decompositions()
+enabled_decompositions: Set[torch._ops.OpOverload] = {
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:14]: If a contributor is writing a converter for one of the `enabled_decompositions`, they should remove it from the set.

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 8]
```diff
@@ -1,10 +1,182 @@
...
+    aten.upsample_bilinear2d,
+    aten.upsample_bilinear2d.vec,
+    aten.upsample_nearest2d_backward,
+    aten.xlogy,
+    aten.zero,
+    aten.zero_,
+    aten.zeros,
+    aten.zeros_like,
+}
+disabled_decompositions: Set[torch._ops.OpOverload] = {}
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:174]: This set is reserved for decompositions that should never appear in the registry or be applied to models.

**Comment by narendasan** [comment]
Have we thought about what this might look like if its user accessible?

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py (hunk 5 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/lowering/_decompositions.py:179]: Would TORCH_DECOMPOSITIONS only include decompositions from the get_torch_decompoistions set?

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py (hunk 7 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/lowering/_decompositions.py:14]: Is there any sort of checking to capture this?

**Comment by narendasan** [comment]
Can we add a tool to monitor these decomposition sets similar to the opset coverage tool?

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py (hunk 5 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/lowering/_decompositions.py:179]: `Note that TORCH_DECOMPOSITIONS may not be a subset of this set`
From what I understand, 
* it seems like some decompositions in `_core_aten_decompositions` have been removed eg: `aten.native_layer_norm`.  [ENABLED_TORCH_DECOMPOSITIONS](https://github.com/pytorch/TensorRT/blob/368a20e99d636b04723fd5e29bdd7d11084434e8/py/torch_tensorrt/dynamo/lowering/_decompositions.py#L14C1-L14C23) is a more complete set (from previous commit maybe ). Is this correct ? 
* In that case, what if we move aten.native.layer_norm to TORCH_TRT_DECOMPOSITIONS since it is useful to us and maybe other useful ones instead of maintaining a `ENABLED_TORCH_DECOMPOSITIONS` which overlaps with `_core_aten_decompositions` one ?

Comment (gs-olive) [reply py/torch_tensorrt/dynamo/lowering/_decompositions.py:179]: @peri044 
- The interpretation of `ENABLED_TORCH_DECOMPOSITIONS` is correct
- My initial intent for `TORCH_TRT_DECOMPOSITIONS` was that it would only store decompositions we specifically (custom) wrote, not ones sourced from Torch, as `layer_norm` would be.

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py (hunk 7 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/lowering/_decompositions.py:14]: Currently, there would only be a runtime check for this. Specifically, the `get_decompositions` function invokes a check that there are no duplicate registrations across the different decomposition registries.

Comment (narendasan) [reply py/torch_tensorrt/dynamo/lowering/_decompositions.py:14]: We might want to think about on registration checking

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 9]
```diff
@@ -1,10 +1,237 @@
...
+        )
+
+    # Conflicts with _core_aten_decompositions will only occur if
+    # enable_experimental_decompositions is True in get_decompositions
+    if aten_op in _core_aten_decompositions:
+        logger.debug(
+            f"Detected custom decomposition for {aten_op}, which conflicts "
+            "with an existing Torch decomposition in core_aten_decompositions. "
+            "The custom implementation will take precedence."
+        )
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:229]: Display warning/info/debug info based on registered decompositions' presence in other existing sets

**Comment by gs-olive** [comment]
The existing opset coverage tool is compatible with this PR, meaning that `get_decompositions` will reflect the degree of coverage of the lowering passes we currently apply via decompositions. It will not show the coverage percentages of using the `core_aten_decompositions`, but this is also not the default usage as of now. We can add in coverage for the `enable_experimental_decompositions` set to the tool as a command line argument, if preferred.

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:1]: Move decomp sets to other file

**Review by peri044 — APPROVED** [review]
LGTM
